HFE.md

```diff
@@ -0,0 +1,643 @@
+2. Introduce a hashtable infrastructure for accessing keys (Preferred)
+    In this approach we suggest to formalize the item access by introducing a new `access` virtual method to the hashtable type, eg:
+    
+
+```
+   typedef enum {
+   ELEMENT_NONE = 0,
+   ELEMENT_VALID,
+   ELEMENT_INVALID,
+   ELEMENT_DELETE,
+} hashtableElementAccessState;
+  
+  typedef struct {
+      const void *(*entryGetKey)(const void *entry);
+      uint64_t (*hashFunction)(const void *key);
+     ...
+      unsigned instant_rehashing : 1;
+     NEW -> hashtableElementAccessState (accessElement)(hashtable ht, void entry);
+ };
+```
+
+Since all items are actually referenced from a hashtable, this will  provide a generic way to acentralize all item access for different object types (eg SETS and ZSETS).
+The main challenge, in this case, is to propagate the expiration context (ie Owning object and database) during hashtable access.
+We propose to include a way to store the current context globally (e.g in TLS), which will require adding set and reset of the context in various command handling places.
+For example:
+
+```
+typedef struct {
+    robj *val;
+    int flags;
+} access_context;
+
+static __thread access_context ht_access_context = {0};
+
+void hashTypeSetAccessContext(robj *o) {
+    ht_access_context.val = o;
+    ht_access_context.flags = 0;
+}
+
+void hashTypeSetAccessContextWithFlags(robj *o, int flags) {
+    ht_access_context.val = o;
+    ht_access_context.flags = flags;
+}
+
+void hashTypeResetAccessContext(void) {
+    ht_access_context.val = NULL;
+    ht_access_context.flags = 0;
+}
+
+// Example use case:
+sds hashTypeGetFromHashTable(robj *o, sds field) {
+    serverAssert(o->encoding == OBJ_ENCODING_HASHTABLE);
+    void *found_element = NULL;
+    hashTypeSetAccessContext(o);
+    hashtableFind(o->ptr, field, &found_element);
+    hashTypeResetAccessContext();
+    if (found_element)
+        return hashTypeEntryGetValue(found_element);
+    else
+        return NULL;
+}
+```
+
+     In this example, the global context is holding the accessed object reference (the database can be taken from the                   server→current_client). It also holds a flags bitmap in order to have better fine grinned control over the item access (for         example avoiding expiring elements during large scan iterations).
+
+Pros:
+     - Minimal code changes.
+     - Less error prone - since this will be applied in every hashtable access we will reduce the risk of missing item being                 accessed in some flows. 
+Cons:
+      - Additional check for `accessElement` existence in the generic hashtable implementation (we have not yet evaluated if
```

Comment:
My intuition is that this will be more degrading than we expect, but it's hard to intuit this without actually seeing the code.

Reply:
Yeh. the draft will be issues this week (if all goes well)